accounting/transferwise/models.py

```diff
@@ -0,0 +1,181 @@
+    auto_create_csv_on_save = models.BooleanField(
+        default=False,
+        help_text=_("Whether this bulk payment should be converted into a CSV file on save."))
+    auto_upload_google_drive_on_save = models.BooleanField(  # pylint: disable=invalid-name
+        default=False,
+        help_text=_("Whether this bulk payment CSV should be uploaded to Google Drive on save."))
+
+    class Meta:
+        verbose_name = _('TransferWise Bulk Payment')
+        verbose_name_plural = _('TransferWise Bulk Payments')
+
+    def __str__(self):
+        """
+        Indicate who's sending this TransferWise bulk payment, and when.
+        """
+        return '{date}: {sender}'.format(date=self.date, sender=self.sender)
+
+    @property
+    def recipients(self):
+        """
+        Return a list of TransferWise recipients for the sender of this bulk payment.
+        """
+        # We get a list of recipients who are connected to the sender and have TransferWise recipient IDs.
+        hourly_rates = self.sender.client_hourly_rates.exclude(provider__bank_accounts__transferwise_recipient_id=None)
+        return [rate.provider for rate in hourly_rates]
+
+    @property
+    def csv_filename(self):
+        """
+        Return a CSV filename for this bulk payment.
+        """
+        return 'transferwise_bulk_payment_csv_{sender}_{date}.csv'.format(
+            sender=self.sender.user.username,
+            date=self.date.strftime("%Y-%m-%d"),
+        )
+
+    def create_payments(self):
+        """
+        Create the individual TransferWise payments needed for this bulk payment.
+
+        The chosen invoice is the one that matches this bulk payment's date's month for the given payment's
+        sender and recipient. If an invoice does not fit this criteria, it gets the latest unpaid one instead.
+
+        Note that the invoice is the source of truth for payments, and thus should exist before making a payment.
+
+        TODO: Besides creating an internal version of the payment for tracking certain data (i.e. recipient ID,
+        TODO: invoice, etc.), this should eventually actually call the API to create real transfers. At that stage,
+        TODO: we could then skip generating CSVs and create all the transfers in one go, and call them a 'bulk payment',
+        TODO: even if that terminology is only to represent how we're doing it internally.
+        """
+        for recipient in self.recipients:
+            try:
+                # Get an invoice matching this bulk payment's month & year.
+                invoice = recipient.provider_invoices.filter(
+                    client=self.sender,
+                    date__year=self.date.year,
+                    date__month=self.date.month,
+                    paid=False,
+                ).latest('date')
+            except Invoice.DoesNotExist:
+                # Fallback to the latest unpaid one, or just skip the recipient.
```

Comment:
Why do we do this? If we create a bulk payment for a given month, it doesn't sound correct to return invoices that are from a different month, even unpaid - that's counter-intuitive and will likely result in mistakes.

Also, what's the use case? A good one to consider could be the one we'll likely face next month with Kshitij, who might end up with both this month and next month's invoices processed in the next batch payment.

To handle this without needing to create two batch payments, could you use a start & end date for the bulk payment, with the start date set by default to an ancient date, and an end date set to now? Then include all unpaid invoices in the batch transfer, even if there are several of them - this way this would by default include all unpaid invoices to date in the transfer, but be explicit about it, and provide a way to override the behavior to select a specific time window.

Reply:
@antoviaque That's a good suggestion to use the start & end dates, yeah. I'll scratch out this invoice gathering code to just get all unpaid invoices in that range then. But we need to be sure, of course, that we're then marking invoices as paid properly. When do we consider an invoice paid, given the current state of the automation?

Reply:
@UmanShahzad I think you can consider an invoice paid once it has been included in a TransferWise bulk payment - for now that should be enough.

Reply:
Done :+1: